### PR TITLE
Interacting with other apps

### PR DESCRIPTION
**Pull request by jonholmberg**
You can now send a message to a number in your contacts! Just choose a recipient from your contacts, write a message and press send! 🐐 

What is your convention for the order of the methods? See for instance `MainFragment.java` - it's a bit messy right now, but I wasn't sure what the best way to order them is? The only thing really consistent is that all the lifecycle methods (`onAttach`, `onCreate` etc) is at the top, which I find quite nice. As an example of my doubt, would you move the helper methods used in `onCreate` up to be just below `onCreate` or leave them be? Hope it makes sense 🥔

### CODE REVIEW

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java [hunk 1]
```diff
@@ -1,26 +1,68 @@
...
+    final static String ENTER_MESSAGE_FRAGMENT_TAG = "EnterMessageFragment";
+
+    // Keys
+    final static String PHONE_NUMBER = "contactPhoneNumber";
+    final static String DISPLAY_NAME = "displayName";
+    final static String CONTACT_DETAILS = "phoneValues";
+    final static String PHONE_DIRECTORY_CONTENT_TYPE = "phoneDirectoryContentType";
+
+    // Map with phone values.
+    private Map<String, String> phoneValues = new HashMap<>();
```
Comment (jonholmberg) [inline jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: I'm blanking a bit - is there a better way to do this? Referring to the second type parameter: they are all taking from `ContactsContract.CommonDataKinds.Phone`.

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java [hunk 2]
```diff
@@ -50,6 +51,8 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
         setupFragments();
+        // Request permission to send SMS.
+        ActivityCompat.requestPermissions(this, new String[]{SEND_SMS}, 1);
```
Comment (jonholmberg) [inline jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:55]: Looking more into this.. 🐴  (Next lesson)

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java [hunk 3]
```diff
@@ -105,12 +210,80 @@ private void setButtonText() {
...
     /** Returns true if the fragment that displays the message is visible. */
     public Boolean isDisplayMessageFragmentVisible() {
         return isDisplayMessageFragmentVisible;
     }
 
+    @Override
+    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
+
+        if (args == null) {
+            return null;
```
Comment (jonholmberg) [inline jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: Don't know what to do here... Log it? 😩

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java (hunk 1 above)
Comment (jonholmberg) [reply jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: Nevermind, just realised, I can just import `ContactsContract.CommonDataKinds.Phone`

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java [hunk 4]
```diff
@@ -62,20 +98,65 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
...
+                }
+
+                delegate.sendMessage(message);
                 editMessage.setText("");
+
+                if (contactPhoneNumber == null) {
+                    return;
+                }
+
+                SmsManager smsManager = SmsManager.getDefault();
```
Comment (tudorgk) [inline jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:157]: SmsManager...

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java [hunk 5]
```diff
@@ -105,12 +193,80 @@ private void setButtonText() {
...
+        data.moveToFirst();
+
+        // Get contact information.
+        int contactNameColumn = data.getColumnIndex(Phone.DISPLAY_NAME);
+        int contactNumberColumn = data.getColumnIndex(Phone.NUMBER);
+
+        String name = data.getString(contactNameColumn);
+        String number = data.getString(contactNumberColumn);
+
+        contactPhoneNumber = number;
```
Comment (tudorgk) [inline jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: Why don't you get the data directly without the aux vars? like:
```java
contactPhoneNumber = data.getString(contactNumberColumn);
contactName = data.getString(contactNameColumn);
```

**Review by tudorgk — COMMENTED** [review]
Looks ok, but `CursorLoaders` are 🤢

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java (hunk 5 above)
Comment (jonholmberg) [reply jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: Don't know. Leftovers I forgot to clean I think. Will do!

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java (hunk 4 above)
Comment (jonholmberg) [reply jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:157]: 😢

**Comment by jonholmberg** [comment]
So should I just merge it?

File: jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java (hunk 3 above)
Comment (jonholmberg) [reply jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: @tudorgk bump: any idea?

Comment (tudorgk) [reply jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java]: Yes, a log here will be nice. Something like `Log.w("TAG", "Loader arguments are null")`

**Comment by tudorgk** [comment]
@jonholmberg When you feel like it, just merge it.

**Pull request merged**